### PR TITLE
Feat/add batch parallelization

### PR DESCRIPTION
**Pull request by conorheins**
# Enhance CVRNN Implementation with Batch Processing and Accuracy Metrics

## Overview
This PR enhances the CVRNN implementation by adding batch processing capabilities and comprehensive accuracy metrics across all example scripts. The changes improve code efficiency and provide better evaluation tools for the segmentation results.

### Core Implementation
- Added type hints using `jaxtyping` for better type safety and documentation
- Updated `CVRNNLayer` and `MultiLayerCVRNN` to use proper type annotations
- Added `PRNGKeyArray` type for better JAX key handling

### Example Scripts
All example scripts (`example_2shapes.py`, `example_3shapes.py`, `example_natural_image.py`) have been updated with:
- Batch processing support via `vmap` for parallel model execution
- Ensemble modeling capabilities with configurable ensemble size
- Comprehensive accuracy metrics using Hungarian matching
- Improved file organization with seed-specific output directories
- Background pixel exclusion option for accuracy computation

### Specific Improvements
1. **Batch Processing**
   - Added `ensemble_size` parameter to run multiple models in parallel
   - Implemented `vmap` for efficient Gaussian sheet initialization
   - Added parallel model creation and execution

2. **Accuracy Metrics**
   - Added Hungarian matching for optimal label assignment
   - Implemented confusion matrix computation
   - Added mean and standard deviation of accuracies across ensemble
   - Added option to exclude background pixels from accuracy computation

3. **File Organization**
   - Created seed-specific output directories
   - Updated file naming to include model index in ensemble
   - Standardized output file structure across all examples

4. **Additional Features**
   - Added `image_index` parameter for 2-shapes and 3-shapes examples
   - Improved command-line argument handling
   - Enhanced documentation and type hints

## Testing
- All example scripts have been tested with various ensemble sizes
- Accuracy metrics have been validated against ground truth
- Batch processing has been verified for correct parallel execution

## Dependencies
Added new dependency:
- `jaxtyping` for improved type safety

## Usage
Example usage with new features:
```bash
# Run with ensemble of 5 models
python example_2shapes.py --ensemble_size 5 --exclude_background

# Run specific image with visualization
python example_3shapes.py --image_index 2 --visualize_dynamics --ensemble_size 3

# Run natural image segmentation with background exclusion
python example_natural_image.py --ensemble_size 4 --exclude_background
```

### CODE REVIEW

File: example_2shapes.py [hunk 1]
```diff
@@ -24,112 +28,169 @@ def main(seed, visualize_dynamics=False):
...
+        X = combined.T                     # (N, nt2)
+        x_full = jnp.angle(X).reshape(Nr, Nc, -1)
+
+        # optional dynamics plot
+        if visualize_dynamics:
+            plot_dynamics(x_full, nt1)
+
+        # spectral clustering
+        x = jnp.exp(1j * x_full.reshape(-1, x_full.shape[2]))
+        valid = ~mask1s[0]
```
[the remark this answers is not in the document]
Comment (conorheins) [reply example_2shapes.py]: Fixed with f42b53a
